frontend/src/features/housingCompany/HousingCompanyImprovementsPage.tsx

```diff
@@ -0,0 +1,239 @@
+    );
+
+    const handleSaveButtonClicked = () => {
+        const formData = {
+            ...housingCompanyData,
+            // Don't send empty improvements to the API
+            improvements: {
+                market_price_index: marketIndexImprovements.filter((i) => i.value) as IImprovement[],
+                construction_price_index: constructionIndexImprovements.filter((i) => i.value) as IImprovement[],
+            },
+        };
+
+        saveHousingCompany({
+            data: formData,
+            id: state?.housingCompany.id,
+        });
+    };
+
+    // Market
+    const handleAddMarketImprovementLine = () => {
+        setMarketIndexImprovements((draft) => {
+            draft.push({
+                key: uuidv4(),
+                name: "",
+                value: null,
+                completion_date: "",
+            });
+        });
+    };
+    const handleSetMarketImprovementLine = (index, fieldPath) => (value) => {
+        setMarketIndexImprovements((draft) => {
+            dotted(draft[index], fieldPath, value);
+        });
+    };
+    const handleRemoveMarketImprovementLine = (index) => {
+        setMarketIndexImprovements((draft) => {
+            draft.splice(index, 1);
+        });
+    };
+
+    // Construction
+    const handleAddConstructionImprovementLine = () => {
+        setConstructionIndexImprovements((draft) => {
+            draft.push({
+                key: uuidv4(),
+                name: "",
+                value: null,
+                completion_date: "",
+            });
+        });
+    };
+    const handleSetConstructionImprovementLine = (index, fieldPath) => (value) => {
+        setConstructionIndexImprovements((draft) => {
+            dotted(draft[index], fieldPath, value);
+        });
+    };
+    const handleRemoveConstructionImprovementLine = (index) => {
+        setConstructionIndexImprovements((draft) => {
+            draft.splice(index, 1);
+        });
+    };
+
+    // Handle saving flow when editing
+    useEffect(() => {
+        if (!isLoading && !error && data && data.id) {
+            hitasToast("Parannukset tallennettu onnistuneesti!");
+            navigate(`/housing-companies/${data.id}`);
+        } else if (error) {
+            hitasToast("Virhe tallentaessa parannuksia!", "error");
+        }
+    }, [isLoading, error, data, navigate]);
+
+    // Redirect user to detail page if state is missing HousingCompany data and user is trying to edit the it
+    // FIXME: Currently does not work, as the page crashes on load due to accessing properties of null
+    useEffect(() => {
+        if (state === null) navigate("..");
+    }, [navigate, state]);
+
+    // FIXME: Rename "ownership" class names for elements when styling
+    return (
+        <div className="view--create view--set-housing-company">
+            <h1 className="main-heading">{housingCompanyData.name.display} Parannukset</h1>
```

Comment:
Minor thingy, but "Parannukset" should probably start with a lower-case first letter :)